### PR TITLE
Removes handling of break, continue and throw from scopes

### PR DESCRIPTION
**Pull request by konradweiss**
This PR removes most of the ScopeManager functionality from the EvaluationOrderGraphPass, only using the functionality to search for labeled statements. Instead of storing break, continue and throws in the scopes. The pass now uses internal maps to save those associations. Additionally the pass renames, refactors and documents parts for better readability and usage.

### CODE REVIEW

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt [hunk 1]
```diff
@@ -467,72 +485,11 @@ class ScopeManager : ScopeProvider {
...
-     * This function SHOULD only be used by the
-     * [de.fraunhofer.aisec.cpg.passes.EvaluationOrderGraphPass] while building up the EOG. It adds
-     * a [LabelStatement] to the list of label statements of the current scope.
-     */
-    fun addLabelStatement(labelStatement: LabelStatement) {
-        currentScope?.addLabelStatement(labelStatement)
-    }
-
     /**
      * This function is internal to the scope manager and primarily used by [addBreakStatement] and
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt:489]: the doc says its internal to the scope manager, but now its public

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.kt [hunk 2]
```diff
@@ -994,42 +1018,41 @@ open class EvaluationOrderGraphPass(ctx: TranslationContext) : TranslationUnitPa
...
     private fun handleLookupScopeStatement(stmt: LookupScopeStatement) {
         // Include the node as part of the EOG itself, but we do not need to go into any children or
         // properties here
-        pushToEOG(stmt)
+        attachToEOG(stmt)
+    }
+
+    /** We use the scope where the current [node] is in, to find a statement labeled with [label] */
+    fun getLabeledASTNode(node: Node, label: String): Node? {
+        scopeManager.enterScopeOfNearestParent(node)
```
[the remark this answers is not in the document]
Comment (konradweiss) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.kt:1050]: Replaced with it

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt (hunk 1 above)
Comment (konradweiss) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt:489]: Done

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.kt [hunk 3]
```diff
@@ -554,18 +555,24 @@ open class EvaluationOrderGraphPass(ctx: TranslationContext) : TranslationUnitPa
...
-            catchingScope.catchesOrRelays[throwType] = currentPredecessors.toMutableList()
+        attachToEOG(node)
+        // Here we identify the encapsulating ast node that can handle or relay a throw
+        val catchingParent =
+            firstParentOrNull(node) { parent ->
+                parent is TryStatement || parent is FunctionDeclaration
+            }
+        if (catchingParent != null) {
+            val throwByTypeMap = nodesToInternalThrows.getOrPut(catchingParent) { mutableMapOf() }
+            val throws = throwByTypeMap.getOrPut(throwType) { mutableListOf() }
```
[the remark this answers is not in the document]
Comment (konradweiss) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.kt]: "throws" are a list of throw statements, in our case UnaryOperators, that are saved under the type they are throwing. Nested tries and the function root can be imagined as a stack. If a try does not implement the necessary catches for an Exception, considering casting, the `handleTryStatement` relays the handling of that Exception to the outer node that is able to handle an Exception. So here we are mainly saving the throws to next associated TryStatement or FunctionDeclaration and when those finish the handling of the necessary children, e.g. catch clauses, the EOG from the catchable throws is attached.

Trying to rename some variables to make that clearer here.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.kt [hunk 4]
```diff
@@ -1095,5 +1118,108 @@ open class EvaluationOrderGraphPass(ctx: TranslationContext) : TranslationUnitPa
...
+                // go up-wards in the ast tree
+                node = node.astParent
+            }
+
+            return null
+        }
+    }
+
+    /**
+     * Statements that constitute the start of the Loop depending on the used pass, mostly of
```
[the remark this answers is not in the document]
Comment (konradweiss) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.kt:1165]: Done

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.kt [hunk 5]
```diff
@@ -1095,5 +1118,108 @@ open class EvaluationOrderGraphPass(ctx: TranslationContext) : TranslationUnitPa
...
+            when (this) {
+                is WhileStatement ->
+                    mutableListOf(this.condition, this.conditionDeclaration).filterNotNull()
+                is ForStatement -> mutableListOf(this.condition).filterNotNull()
+                is ForEachStatement -> mutableListOf(this.variable).filterNotNull()
+                is DoStatement -> mutableListOf(this.condition).filterNotNull()
+                is AssertStatement -> mutableListOf(this.condition).filterNotNull()
+                else -> {
+                    LOGGER.error(
+                        "Currently the component {} does not have defined conditions",
```
[the remark this answers is not in the document]
Comment (konradweiss) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.kt:1216]: Leftover comment when we did not have components, and I carelessly used it as a synonym for arbitrary AST nodes.

Replaced by node type

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.kt [hunk 6]
```diff
@@ -1095,5 +1118,108 @@ open class EvaluationOrderGraphPass(ctx: TranslationContext) : TranslationUnitPa
...
+    /**
+     * Statements that constitute the start of the condition evaluation, mostly of size 1. This has
+     * to be extended if new nodes are added that have a condition relevant as entry points when
+     * looping.
+     */
+    val Node.conditions: List<Node>
+        get() =
+            when (this) {
+                is WhileStatement ->
+                    mutableListOf(this.condition, this.conditionDeclaration).filterNotNull()
```
[the remark this answers is not in the document]
Comment (konradweiss) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.kt:1209]: Done

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt [hunk 7]
```diff
@@ -321,6 +321,24 @@ class ScopeManager : ScopeProvider {
         }
     }
 
+    /**
+     * Enters the scope of the closest astParent that is associated to a scope. This function can be
+     * used to enter the appropriate enclosing scope of any ast node, as long as the frontend build
+     * the scope tree over the entire code base. Entering is conditional, if no parent is scoped,
+     * this function does not create a new scope, similar to [enterScopeIfExists]
+     */
+    fun enterScopeOfNearestParent(node: Node?) {
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt:330]: left over

**Review by oxisto — APPROVED** [review]
LGTM except two things:
- Leftover function
- It seems that the aspect of catching a throw by its type is still not fully implemented with regards to a derived type, unless I don't understand the implementation correctly. If we do not fix this here, we should have at least an issue for that

**Comment by KuechA** [comment]
In #1741 I realized that there's an issue if there's no type of the throw statement and the throw statement has no outgoing EOG edges. We should define a behavior in this case and ideally test it. Maybe it's another PR? Tracked in #1794

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt (hunk 7 above)
Comment (konradweiss) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt]: removed and used `jumpTo` there too

**Comment by konradweiss** [comment]
> In #1741 I realized that there's an issue if there's no type of the throw statement and the throw statement has no outgoing EOG edges. We should define a behavior in this case and ideally test it. Maybe it's another PR? Tracked in #1794

I would solve that in a new PR due to the ThrowStatement being an addition of another PR. This PR was supposed to be a fix of a specific problem that required the removal of scope usage in the pass. If we want to extend or fix functional behavior I would see that in another PR.

**Pull request merged**